crates/pipeline_manager/src/db/mod.rs

```diff
@@ -717,8 +989,154 @@ impl Storage for ProjectDB {
         }
     }
 
+    /// Version the current pipeline object and all state reachable from it.
+    ///
+    /// We store the last revision number in the pipeline object itself.
+    async fn create_pipeline_revision(
+        &self,
+        tenant_id: TenantId,
+        pipeline_id: PipelineId,
+    ) -> AnyResult<Revision> {
+        let mut client = self.pool.get().await?;
+        let txn = client.transaction().await?;
+
+        // TODO: Ideally these queries happen in the same transaction `txn`.
+        // Need to revise the API a bit to optionally pass in a transaction
+        // object for this to work seamlessly. Also probably the program row
+        // should be locked with FOR UPDATE while reading to prevent the
+        // compiler from resetting the status field.
+        let (_pipeline, program, _connectors) =
+            self.pipeline_is_committable(tenant_id, pipeline_id).await?;
+
+        // Find the revision number
+        let revision_data = txn
+            .query_opt(
+                "SELECT last_revision FROM pipeline WHERE id = $1 AND tenant_id = $2 FOR UPDATE",
+                &[&pipeline_id.0, &tenant_id.0],
+            )
+            .await?;
+        let (revision, prev_revision) = match revision_data {
+            Some(row) => {
+                let revision: i64 = row.get::<_, Option<i64>>(0).unwrap_or(0);
+                (revision + 1, revision)
+            }
+            None => return Err(DBError::UnknownPipeline(pipeline_id).into()),
+        };
+
+        // Check if we actually changed something before writing a new revision
+        //
+        // Note: What fields are checked is ultimately determined by whatever is
+        // used by the pipeline configuration, e.g., what the `start` function
+        // uses in `runner.rs` to write the config/metadata:
+        let change = txn.query_one(
+            "SELECT count(*) FROM
```

Comment:
SQL strings.... is this the state-of-the-art?

Reply:
This is basically a prepared statement so we are not constructing query strings dynamically. That said, Rust doesn't have good tooling like jOOQ in java to write type-safe SQL, so this is the best we can do for now.